File: strings/open_the_lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> deadend(deadends.begin(), deadends.end());
        if (deadend.contains(target)) return -1;

        string starter = "0000";
        if (deadend.contains(starter)) return -1;

        vector<int> moves = {-1, 1};
        queue<pair<string, int>> q;
        q.push({starter, 0});

        unordered_set<string> visited;
        visited.insert(starter);

        while (!q.empty()) {
            auto [currentCombo, actions] = q.front();
            q.pop();

            if (currentCombo == target) return actions;

            for (int i = 0; i < 4; ++i) {
                for (int move : moves) {
                    int newDigit = (currentCombo[i] - '0' + move + 10) % 10;
                    string newCombo = currentCombo;
                    newCombo[i] = newDigit + '0';

                    if (!deadend.contains(newCombo) && !visited.contains(newCombo)) {
                        visited.insert(newCombo);
                        q.push({newCombo, actions + 1});
                    }
                }
            }
        }

        return -1;
    }
};

File: strings/open_the_lock.cs
public class Solution {
    public int OpenLock(string[] deadends, string target) {
        HashSet<string> deadendsSet = [..deadends];
        if (deadendsSet.Contains(target)) return -1;
        
        string starter = "0000";
        if (deadendsSet.Contains(starter)) return -1;

        Span<int> moves = [-1, 1];
        Queue<(string, int)> q = [];
        q.Enqueue((starter, 0));

        HashSet<string> visited = [starter];

        try {
            while (q.TryDequeue(out var current)) {
                var (combination, actions) = current;
                if (combination == target) return actions;

                for (int i = 0; i < combination.Length; ++i) {
                    foreach (int move in moves) {
                        StringBuilder sb = new(combination);
                        sb[i] = (char) (((combination[i] - '0' + move + 10) % 10) + '0');
                        string newCombination = sb.ToString();

                        if (!deadendsSet.Contains(newCombination) && visited.Add(newCombination)) {
                            q.Enqueue((newCombination, actions + 1));
                        }
                    }
                }
            }

            return -1;
        } finally {
            GC.Collect(0);
        }
    }
}
